hera_qm/vis_metrics.py

```diff
@@ -48,6 +51,138 @@ def check_noise_variance(data):
     return Cij
 
 
+def sequential_diff(data, t_int=None, axis=(0,), pad=False, add_to_history=''):
+    """
+    Take a sequential (forward) difference of a visibility waterfall
+    as an estimate of the visibility noise. Note: the 1/sqrt(2) correction
+    needed after taking a difference is not applied to the data directly,
+    but is taken into account by the output t_int--or if data is a UVData
+    object, multiplied into the nsample_array.
+
+    Parameters
+    ----------
+    data : ndarray or UVData object
+        A 2D or 3D ndarray containing visibility data, with
+        shape (Ntimes, Nfreqs, :). Or UVData object
+        holding visibility data.
+
+    t_int : ndarray
+        A 2D or 3D ndarray containing the integration time
+        of the visibilities matching shape of data.
+        If None and data is UVData, will use UVData ~flags *
+        nsample * integration_time as input.
+
+    axis : int or tuple
+        Axes along which to take sequential difference
+
+    pad : boolean
+        If True, insert an extra (flagged) column at the end
+        of axis such that diff_data has same shape as input.
+
+    add_to_history : str
+        A string to add to history of UVData if provided, in addition
+        to standard history comment.
+    
+    Returns
+    -------
+    diff_data : ndarray or UVData object
+        A 2D or 3D complex differenced data array,
+        or UVData object holding differenced data
+        divided by sqrt(2).
+
+    diff_t_int : ndarray
+        If input data is ndarray, this is the inverse
+        sum of t_ints in the difference.
+    """
+    # type check
+    if isinstance(axis, (int, np.integer)):
+        axis = (axis,)
+    if not np.all([ax in [0, 1] for ax in axis]):
+        raise ValueError("axis must be 0, 1 or both")
+
+    # perform differencing if data is an ndarray
+    if isinstance(data, np.ndarray):
+        # get t_int
+        if t_int is None:
+            t_int = np.ones_like(data, dtype=np.float64)
+        else:
+            if not isinstance(t_int, np.ndarray):
+                raise ValueError("t_int must be ndarray if data is ndarray")
+
+        # pad
+        if pad:
+            for ax in axis:
+                # get padding vector
+                p = utils.dynamic_slice(np.zeros_like(data, dtype=np.float), slice(0, 1), axis=ax)
+                # pad arrays
+                data = np.concatenate([data, p], axis=ax)
+                t_int = np.concatenate([t_int, p], axis=ax)
+
+        # difference data
+        for ax in axis:
+            data = (utils.dynamic_slice(data, slice(1, None), axis=ax) \
+                   - utils.dynamic_slice(data, slice(None, -1), axis=ax))
+            t_int = 1.0/(1.0/utils.dynamic_slice(t_int, slice(1, None), axis=ax) \
+                         + 1.0/utils.dynamic_slice(t_int, slice(None, -1), axis=ax))
```

Comment:
I don't have a problem w/ dividebyzero warning here, b/c that is indeed what I want it to do: the 1./0. will give me an inf and the 1/inf will then give me a zero, which is what I want.

If I use the `true_divide` then the zeros are left untouched, and that means when I do the sum against another array which might not be zero, then it won't do what I want.

For example, if
```
arr1 = [1.0, 1.0, 0.0]
arr2 = [1.0, 1.0, 1.0]
```
I want the inverse average of these to be
```
inv_avg = [0.5, 0.5, 0.0]
```
But in this case if I just do
```
s = np.true_divide(1, arr1, where=~np.isclose(arr1, 0.0)) + np.true_divide(1, arr2, where=~np.isclose(arr2, 0.0))
t_int = np.true_divide(1, s, where=~np.isclose(s, 0.0))
```
then I end up with
```
[0.5, 0.5, 1. ]
```

Reply:
Let's work on this. Using divide by zeros can be dangerous, and because of how python reports warnings, it could potentially hide other warnings elsewhere (each warning is only reported once). Given what you wrote above, I think the following will do what you want:
```s = np.true_divide(1, arr1, where=~np.isclose(arr1, 0.0)) + np.true_divide(1, arr2, where=~np.isclose(arr2, 0.0))
t_int = np.true_divide(1, s, where=~np.isclose(s, 0.0))
t_int = np.where(0, t_int, np.isclose(s, 0.0))
```
@plaplant I know you've dealt with stuff like this before, do you have a more elegant idea?

Reply:
@adampbeardsley I 100% agree that we should avoid dividing by zero as much as possible. I don't have a more elegant solution than that, but I think `np.where` is relatively cheap, so it shouldn't add too much overhead.